c10/cuda/CUDAException.h

```diff
@@ -81,5 +94,152 @@ C10_CUDA_API void c10_cuda_check_implementation(
+  std::string launch_stacktrace;
+  /// Kernel that was launched
+  std::string kernel_name;
+  /// Device the kernel was launched on
+  int device;
+  /// Stream the kernel was launched on
+  int32_t stream;
+  /// A number that uniquely identifies the kernel launch
+  uint32_t generation_number;
+};
+
+/// Circular buffer used to hold information about kernel launches
+/// this is later used to reconstruct how a device-side kernel assertion failure
+/// occurred CUDAKernelLaunchRegistry is used as a singleton
+class C10_CUDA_API CUDAKernelLaunchRegistry {
+ private:
+  /// Assume that this is the max number of kernel launches that might ever be
+  /// enqueued across all streams on a single device
+  static constexpr int max_kernel_launches = 1024;
+  /// Void deleter for UVM/managed memory pointers
+  static void uvm_deleter(DeviceAssertionsData* uvm_assertions_ptr);
+  /// How many kernel launch infos we've inserted. Used to ensure that circular
+  /// queue doesn't provide false information by always increasing, but also to
+  /// mark where we are inserting into the queue
+  uint32_t generation_number = 0;
+  /// Shared mutex between writer and accessor to ensure multi-threaded safety.
+  mutable std::mutex read_write_mutex;
+  /// Pointer to managed memory keeping track of device-side assertions. There
+  /// is one entry for each possible device the process might work with. Unused
+  /// entries are nullptrs. We could also use an unordered_set here, but this
+  /// vector design will be faster and the wasted memory is small since we
+  /// expect the number of GPUs per node will always be small
+  std::vector<
+      std::unique_ptr<DeviceAssertionsData, void (*)(DeviceAssertionsData*)>>
+      uvm_assertions;
+  /// A single circular buffer holds information about every kernel launch the
+  /// process makes across all devices.
+  std::vector<CUDAKernelLaunchInfo> kernel_launches;
+  /// Get the number of CUDA devices
+  static int get_device_count();
+  /// Get current device id - a little redundant, but helps avoid a circular
+  /// include
+  static int get_device_id();
+  /// Get current device compute capability - note that this dangerously assumes
+  /// that if one CUDA GPU supports device-side assertions they all do. This is
+  /// probably fine since the latest CUDA GPU that doesn't support UVM is the
+  /// K80 released 2014-11-17. Mixing that GPU with a newer one is likely to be
+  /// rare enough that the defensive
+  static int get_device_compute_capability(const int device_num);
+  static bool check_if_all_devices_support_managed_memory();
+  const bool do_all_devices_support_managed_memory = false;
+  bool check_env_for_enable_launch_stacktracing() const;
+  bool check_env_for_dsa_enabled() const;
+
+ public:
+  CUDAKernelLaunchRegistry();
+  /// Register a new kernel launch and obtain a generation number back to be
+  /// passed to the kernel
+  uint32_t insert(
+      const std::string& launch_filename,
```

Comment:
It seems unnecessary for these to be std::string; you're always passing in `__FILE__` so these can just be `const char*`. It seems good to apply this optimization, as you are ALWAYS populating these fields for every DSA cuda launch, seems much better to not have to do a dynamic allocation every time you do this.

Reply:
Good call.